torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,244 @@
+    # the output of the module, for input QuantStub will observe them
+    if hasattr(module, 'qconfig') and module.qconfig is not None and len(module._modules) == 0:
+        # observer and hook will be gone after we swap the module
+        module.add_module('observer', module.qconfig.activation())
+        module.register_forward_hook(_observer_forward_hook)
+
+class QuantWrapper(nn.Module):
+    r"""A wrapper class that wraps the input module, adds QuantStub and
+    DeQuantStub and surround the call to module with call to quant and dequant
+    modules.
+
+    This is used by the `quantization` utility functions to add the quant and
+    dequant modules, before `convert` function `QuantStub` will just be observer,
+    it observes the input tensor, after `convert`, `QuantStub`
+    will be swapped to `nnq.Quantize` which does actual quantization. Similarly
+    for `DeQuantStub`.
+    """
+    def __init__(self, module):
+        super(QuantWrapper, self).__init__()
+        qconfig = module.qconfig if hasattr(module, 'qconfig') else None
+        self.quant = QuantStub(qconfig)
+        self.dequant = DeQuantStub()
+        self.module = module
+
+    def forward(self, X):
+        X = self.quant(X)
+        X = self.module(X)
+        return self.dequant(X)
+
+def add_quant_dequant(module):
+    r"""Wrap the leaf child module in QuantWrapper if it has a valid qconfig
+    Note that this function will modify the children of module inplace and it
+    can return a new module which wraps the input module as well.
+
+    Args:
+        module: input module with qconfig attributes for all the leaf modules
+        that we want to quantize
+
+    Return:
+        Either the inplace modified module with submodules wrapped in
+        `QuantWrapper` based on qconfig or a new `QuantWrapper` module which
+        wraps the input module, the latter case only happens when the input
+        module is a leaf module and we want to quantize it.
+    """
+    if len(module._modules) == 0 and hasattr(module, 'qconfig') and module.qconfig:
+        return QuantWrapper(module)
+
+    for name, child in module.named_children():
+        module._modules[name] = add_quant_dequant(child)
+    return module
+
+def prepare(module, qconfig_dict):
+    r"""Prepares the module for calibration or training given a qconfig_dict.
+    Note that the module will be modified inplace but in case the input module
+    is a leaf module, a wrapped module will be returned.
+
+    Args:
+        mod: input module
+        qconfig_dict: dictionary that maps from name of submodule to quantization
+                      configuration
+    Return:
+        A module with qconfig propogated, observer and quant dequant or fake
+        quant modules attached, a module that is ready for calibration or
+        training
+    """
+    propagate_qconfig(module, qconfig_dict)
+    module = add_quant_dequant(module)
+    add_observer(module)
+    return module
+
```

Comment:
Make add_quant_dequant optional: This allows for quantize (the top level API) to work for cases where the user manually inserts quantstubs or calls quantwrapper.
```
def prepare(module, qconfig_dict, enable_add_quant_dequant=True):
    propagate_qconfig(module, qconfig_dict)
   if enable_add_quant_dequant:
        module = add_quant_dequant(module)
    add_observer(module)
```
    return module

Reply:
moving add_quant_dequant as discussed.